src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -0,0 +1,477 @@
+    {
+        return Optional.empty();
+    }
+
+    @Override
+    public boolean dependsOn(ColumnMetadata column)
+    {
+        return indexContext.getDefinition().compareTo(column) == 0;
+    }
+
+    @Override
+    public boolean supportsExpression(ColumnMetadata column, Operator operator)
+    {
+        return dependsOn(column) && indexContext.supports(operator);
+    }
+
+    @Override
+    public AbstractType<?> customExpressionValueType()
+    {
+        return null;
+    }
+
+    @Override
+    public RowFilter getPostIndexQueryFilter(RowFilter filter)
+    {
+        // it should be executed from the SAI query plan, this is only used by the singleton index query plan
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long getEstimatedResultRows()
+    {
+        throw new UnsupportedOperationException("Use StorageAttachedIndexQueryPlan#getEstimatedResultRows() instead.");
+    }
+
+    @Override
+    public boolean isQueryable(Status status)
+    {
+        // consider unknown status as queryable, because gossip may not be up-to-date for newly joining nodes.
+        return status == Status.BUILD_SUCCEEDED || status == Status.UNKNOWN;
+    }
+
+    @Override
+    public void validate(PartitionUpdate update) throws InvalidRequestException
+    {}
+
+    private class UpdateIndexer extends IndexerAdapter
+    {
+        private final DecoratedKey key;
+        private final Memtable mt;
+        private final WriteContext writeContext;
+
+        UpdateIndexer(DecoratedKey key, Memtable mt, WriteContext writeContext)
+        {
+            this.key = key;
+            this.mt = mt;
+            this.writeContext = writeContext;
+        }
+
+        @Override
+        public void insertRow(Row row)
+        {
+            adjustMemtableSize(indexContext.index(key, row, mt), CassandraWriteContext.fromContext(writeContext).getGroup());
+        }
+
+        @Override
+        public void updateRow(Row oldRow, Row newRow)
+        {
+            insertRow(newRow);
+        }
+
+        void adjustMemtableSize(long additionalSpace, OpOrder.Group opGroup)
+        {
+            mt.markExtraOnHeapUsed(additionalSpace, opGroup);
+        }
+    }
+
+    protected static abstract class IndexerAdapter implements Indexer
+    {
+        @Override
+        public void begin() { }
+
+        @Override
+        public void finish() { }
+
+        @Override
+        public void partitionDelete(DeletionTime dt)
+        {
+        }
+
+        @Override
+        public void rangeTombstone(RangeTombstone rt)
+        {
+        }
+
+        @Override
+        public void removeRow(Row row)
+        {
+        }
+    }
+
+    @Override
+    public Searcher searcherFor(ReadCommand command) throws InvalidRequestException
+    {
+        // searchers should be created from the query plan, this is only used by the singleton index query plan
+        throw new UnsupportedOperationException();
```

Comment:
nit: How about just putting the comment (or at least the first part of it) in the exception message?